### ISSUE TITLE
Completion of contract campaigns in last few hours

### ISSUE THREAD

**Issue by erikgeurts** [revive-adserver/revive-adserver#319]
Contract campaigns have a fixed duration, for example 7 days or 31 days. They have a target to deliver a specific number of impressions, for example 1 million.

Often, contract campaigns have trouble reaching their target, because the number of impressions in the final few hours of the last day of the campaign's duration are lower then they should have been. In a short campaign, this can result in under-delivery of a few percent of the total, a situation that might be avoidable when a common cause I will explain below can be mitigated.

In many cases, this form of structural under-delivery can be explained, because early evening traffic on the zones where the campaign runs, is at a normal level, for example between 7PM and 10 PM. However, human beings are creatures of habit, so a large portion of the population will go to sleep some time between the hours of 10PM and midnight. As a result, impressions between 10 and 11 PM, and then again between 11 PM and midnight, are lower than forecasted, since the forecast is calculated by the hourly maintenance process based on the previous hour's statistics.

The consequence is that a contract campaign will, in many cases, fail to meet its target because the maintenance process sets the probability too low. It is expecting more impressions during the last few hours of the day than what is going to happen for real. In relative terms, the effect of this misalignment is bigger with campaigns that have a short duration (like a few days or a single week).

Before version 2.8, the maintenance process used to look back not just at the previous hour be in fact many days. If I'm not mistaken, this has been abandoned because of scalability issues.

I would like to propose that we attempt to build in an adjustment model for the forecasting algorithm based on a self-learning and self-adjusting 'typical 24 hour day pattern', or perhaps even a 'typical 168 hour week pattern'.

One complicating factor in this proposal is that many installations of Revive Adservers are used in geographical areas that have multiple time zones. I would welcome suggestions by others who are more familiar with this type of math and statistics.

**Comment by rhapsodyv** [comment]
Hi,
I'm having a similar issue here. I have contract campaigns cpm limited by daily impressions and without end date.
I was using UTC -3 in default account preferences. The campaigns never reach their limits.
But, in this weekend, I changed to UTC 0 and now every contract campaing is working very very fine and reaching theirs daily limit.
I'm analyzing the source code, but I suspect that in the last our the required impressions just don't is set to the remaining when the campaing isn't in UTC 0.

@mbeccati

**Comment by erikgeurts** [comment]
@rhapsodyv Can you confirm that you're running the newest v3.1.0 that has this fix for time zone in it?

**Comment by rhapsodyv** [comment]
Yes. I'm using version 3.1.0. We upgraded our servers two weeks ago. I tried use overdeliver to 200%, but neither way the campaigns reach their target.
Setting default account to UTC 0 solved the problem even with overdeliver off.

**Comment by erikgeurts** [comment]
OK, thanks for confirming. This is must unexpected. We would welcome any thoughts you might have on this, because so far in all of our testing, this tz issue was fixed in v3.1.0.

**Comment by rhapsodyv** [comment]
A clue: My early debugging points that the problem is with UTC negative. With UTC negative, the end day for "today" goes to tomorow when transformed to UTC 0.
For example:
At UTC -3:
start date will be 00:00 UTC -3 => 03:00 UTC 0
end date will be 23:59 UTC -3 => 02:59 UTC 0 (of tomorow)
And I suspect the problem occur when calculate the require_impressions when "now" is 21:00 UTC -3, that "today" for the campaign is already "tomorow" in UTC 0.

But it's just my suspects.

**Comment by mbeccati** [comment]
@rhapsodyv @erikgeurts sure, there's a chance that I missed something during my recent revivew and fix, once the scenarios I was looking at were working as expected.

**Comment by rhapsodyv** [comment]
I think I found....

The method: 
```php
OA_Maintenance_Priority_AdServer_Task_GetRequiredAdImpressions::getCampaignImpressionInventoryRequirement
```

In this part:
```php 
                $oTodayDate = new Date($this->_getDate());
                $oTodayDate->convertTZ($this->currentTz);
                $oCampaign->setSummaryStatisticsToday($oTodayDate->format('%Y-%m-%d'));
```
This code query campaign delivered statistics for today, but it's not considering the start and end time for the campaing timezone.

The query executed takes Today and filters for 00:00 to 23:59:59 UTC 0, but a campaing with UTC -3 must filter for: 03:00 to 02:59:59 UTC 0.
```php
        array($joinTable2, "$joinTable1.bannerid = $joinTable2.ad_id AND $joinTable2.date_time >= '$today 00:00:00' AND $joinTable2.date_time <= '$today 23:59:59'")
```
Full query:
```php
    function getCampaignDeliveryToday($id, $today)
    {
        $aConf = $GLOBALS['_MAX']['CONF'];
        $query = array();
        $table = $this->_getTablenameUnquoted('campaigns');
        $joinTable1 = $this->_getTablenameUnquoted('banners');
        $joinTable2 = $this->_getTablenameUnquoted('data_intermediate_ad');

        $query['table']    = $table;
        $query['fields']   = array(
                                "SUM($joinTable2.requests) AS sum_requests",
                                "SUM($joinTable2.impressions) AS sum_views",
                                "SUM($joinTable2.clicks) AS sum_clicks",
                                "SUM($joinTable2.conversions) AS sum_conversions",
                                "$table.campaignid AS placement_id"
        );
        $query['wheres']   = array(
        array("$table.campaignid = $id", 'AND')
        );
        $query['joins']    = array(
        array($joinTable1, "$table.campaignid = $joinTable1.campaignid"),
        array($joinTable2, "$joinTable1.bannerid = $joinTable2.ad_id AND $joinTable2.date_time >= '$today 00:00:00' AND $joinTable2.date_time <= '$today 23:59:59'")
        );
        $query['group']    = "placement_id";
        return $this->_get($query);
    }
```

The code use it to calculate the remaining impressions for today. But it just don't takes in account the time of the start and end dates, that isn't 00:00 and 23:59 for campaigns with UTC negative or positive.

I will do more tests.

**Comment by mbeccati** [comment]
@rhapsodyv Nice find, thanks!

Would you mind opening a new issue and posting your findings? I'm going to close this one to avoid ending up with a giantic thread on a bug that was closed in 3.1.0.

**Comment by rhapsodyv** [comment]
Sure! New issue: https://github.com/revive-adserver/revive-adserver/issues/497
Thanks!!
